### PR TITLE
SearchBar Refactor & Redesign

### LINKED ISSUE #3360: Refactor: SearchBar component
too many lines for nothing

### LINKED ISSUE #3782: Search while write doesn't work
We've recently added Indices to make the search work while you typing fast.

https://github.com/kodadot/rubick/issues/123#event-7177423257

but seems it doesn't work

Might it be resolved in #3360 ?

### Testing cases
- Spirit Key
- Skullz Art

image

image

image

OR it works on some terms only? why 👀

image

### LINKED ISSUE #3803: option to reset search terms
#### Step to reproduce
- go to beta
- search "test"
- now I want to reset my query
- erase "test" & press enter
- still looking at "test" results

#### possible solution
- let user make query even when there is no text

could be linked with #3360

### LINKED ISSUE #3843: Add skeleton with 3-5 items in search bar
We should show a skeleton items with at least 3-5 items in collections and NFTs till the client sees the final results 
This should be done in 
- #3763 

https://user-images.githubusercontent.com/5887929/186915387-62dfe8c7-451b-47cf-a0cb-fcebcb679d38.mov

inspiration from looksrare ^

new search on slow 3G network  
- #3763 

https://user-images.githubusercontent.com/5887929/186916046-9c7f4a54-8b89-4f28-94e9-e2953880b674.mov

### LINKED ISSUE #3858: Remove extra line at search at mobile
Let's remove small input and keep only the full-width one. 

https://user-images.githubusercontent.com/5887929/187070049-ecae64ee-5ebf-45df-8923-cbb49f5d0b55.mov

@Jarsen136 could do it in #3763

### PR DESCRIPTION
**Pull request by Jarsen136**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [ ] Bugfix
- [x] Feature
- [x] Refactoring

### What's new?

- [x] PR closes #3360
- [x] PR closes #3782
- [x] PR closes #3803
- [x] PR closes #3843
- [x]  PR closes #3858

### Before submitting Pull Request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [x] I've posted a screenshot of demonstrated change in this PR

### Optional

- [ ] I've tested it at </rmrk/collection/26902bc2f7c20c546a-1FVG7>
- [ ] I've tested PR on mobile and everything seems works
- [ ] I found edge cases
- [ ] I've written some unit tests 🧪

### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG)

### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

### Screenshot

- [ ] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

### CODE REVIEW

**Comment by DeluneClaire** [comment]
Unfortunately,
Snímek obrazovky 2022-08-18 v 12 59 37
 I can't see the stuff that was in the mentioned PR...

**Comment by Jarsen136** [comment]
> Unfortunately,
> 
>  I can't see the stuff that was in the mentioned PR...

It is not ready for review. Still work in progress

**Comment by yangwao** [comment]
Thinking of adding keyboard shortcuts flow like:

If I write term, like "Skullz" 
I would like to with down arrow go and select item
with enter open that item

other than that, really great work there! :)
like it a lot!

**Comment by yangwao** [comment]
I've noticed scroll on results 👀
i'll suggest removing it to make it seamless and fixed for now 

https://user-images.githubusercontent.com/5887929/185588032-52f612c3-cd9f-45fe-89e9-c3aef50260ab.mov

**Comment by Jarsen136** [comment]
✅  keyboard shortcuts
✅  remove scroll on results
✅  hide search result on mobile

**Comment by yangwao** [comment]
Maybe the white text on black is missing? 

image

image

image

@vikiival is working on supplemental PR, which replaces using SubQuery and leverages SubSquid and it will be actually way faster

**Comment by Jarsen136** [comment]
> Maybe the white text on black is missing?
> 

Will check it

**Comment by yangwao** [comment]
- #3817 👀

**Comment by Jarsen136** [comment]
I think its functionality is ready now on both rmrk & bsx chains.

File: components/rmrk/Gallery/Search/SearchBar.vue [hunk 1]
```diff
@@ -170,12 +237,18 @@
 </template>
 
 <script lang="ts">
-import { Component, Emit, Prop, Ref, mixins } from 'nuxt-property-decorator'
+import {
```
Comment (preschian) [inline components/rmrk/Gallery/Search/SearchBar.vue:240]: hi @Jarsen136, is it possible to move `Search` components outside `rmrk` folders? maybe move the component to `/components/search/`

File: components/rmrk/Gallery/Search/SearchBar.vue [hunk 2]
```diff
@@ -235,18 +306,24 @@ export default class SearchBar extends mixins(
...
     number | string | undefined
   ] = [undefined, undefined]
   private sliderDirty = false
-  private searchSuggestionEachTypeMaxNum = 3
+  private searchSuggestionEachTypeMaxNum = 5
   private bigNum = 1e10
   private keyDownNativeEnterFlag = true
-  private defaultNFTSuggestions: (NFTWithMeta & { itemType: 'NFT' })[] = []
-  private defaultCollectionSuggestions: CollectionWithMeta[] = []
+  public defaultCollectionSuggestions: CollectionWithMeta[] = []
```
Comment (roiLeo) [inline components/rmrk/Gallery/Search/SearchBar.vue:319]: `CollectionWithMeta` missing property `total` & `uniqueCollectors`

File: components/rmrk/Gallery/Search/SearchBar.vue [hunk 3]
```diff
@@ -21,74 +21,141 @@
...
-          <template #footer>
-            <a
-              v-if="autocompleteFooterShow"
-              class="navbar-item"
-              style="justify-content: center; margin-right: 0.8em"
-              @click.stop.prevent="searchSuggestionEachTypeMaxNum = bigNum">
-              All results
-            </a>
+                <nuxt-link :to="{ name: 'series-insight' }">
+                  <div>Rankings --></div>
```
Comment (roiLeo) [inline components/rmrk/Gallery/Search/SearchBar.vue:155]: missing `Rankings` tranlation

File: components/rmrk/Gallery/Search/SearchBar.vue [hunk 4]
```diff
@@ -21,74 +21,141 @@
...
+                  </div>
                 </div>
-                <div class="media-content">
-                  {{ props.option.name }}
+                <nuxt-link
+                  :to="{
+                    name: routeOf('explore'),
+                    query: { ...$route.query, tab: 'GALLERY' },
+                  }">
+                  <div>See All --></div>
```
Comment (roiLeo) [inline components/rmrk/Gallery/Search/SearchBar.vue:115]:
```suggestion
                  <div>See All --></div>
```
missing `See All` translation

File: components/rmrk/Gallery/Search/SearchBar.vue [hunk 5]
```diff
@@ -21,74 +21,141 @@
...
+                      <div
+                        class="is-flex is-flex-direction-row is-justify-content-space-between pt-1 pr-2">
+                        <span class="name">{{ item.collection?.name }}</span>
+                        <span>
+                          Price:
+                          <Money
+                            v-if="item.price && parseInt(item.price)"
+                            :value="item.price"
+                            inline /><span v-else>None</span>
+                        </span>
```
Comment (roiLeo) [inline components/rmrk/Gallery/Search/SearchBar.vue:105]: hmmm... 🤔

**Review by roiLeo — COMMENTED** [review]
Some comments:
- could we replace `round-image` class with `rounded` prop
- I would remove "Price: None" when item is not Listed
- refactor/create new media component (reusable)
- no more history tab?

File: components/rmrk/Gallery/Search/SearchBar.vue (hunk 1 above)
Comment (yangwao) [reply components/rmrk/Gallery/Search/SearchBar.vue:240]: yes! For future multi-chain components it would be really appreciated :)

**Comment by yangwao** [comment]
just noticed update toast has changed colours here ?
image

**Comment by yangwao** [comment]
- Possibility to select all rankings/See All (at collection & NFTs tab) with arrow
- Moving text when selecting seems to be happening on Trending, Collections and NFTs ? 
https://user-images.githubusercontent.com/5887929/186630129-bf56d661-33eb-48ca-a596-9a8b736e56b6.mov

Really great progress, I like it! :)

**Comment by Jarsen136** [comment]
Thanks a lot. Will check these comments later.

File: components/rmrk/Gallery/Search/SearchBar.vue (hunk 1 above)
Comment (Jarsen136) [reply components/rmrk/Gallery/Search/SearchBar.vue:240]: Done

File: components/rmrk/Gallery/Search/SearchBar.vue (hunk 5 above)
Comment (Jarsen136) [reply components/rmrk/Gallery/Search/SearchBar.vue]: Updated

Comment (roiLeo) [reply components/rmrk/Gallery/Search/SearchBar.vue]: can you please use "price" from en.json?

**Comment by yangwao** [comment]
Made issue to add skeleton items to the search 
- https://github.com/kodadot/nft-gallery/issues/3843

**Comment by yangwao** [comment]
Found two bugs; maybe you are aware of them 😄

When I'm in the collection and search for "sunset" and select it with keyboard, dialogue should disappear 

https://user-images.githubusercontent.com/5887929/186918700-255e7064-6ef5-45a4-a7a7-149c9c2e3f09.mov

Another is quite weird; when I start from landing and write "sunset" and select collection Dog sunset, it opens the trending collection Kanaria

https://user-images.githubusercontent.com/5887929/186919024-3beaba24-a4c5-46ab-8873-352650bc7a61.mov

File: components/rmrk/Gallery/Search/SearchBar.vue (hunk 5 above)
Comment (Jarsen136) [reply components/rmrk/Gallery/Search/SearchBar.vue]: Yes, already updated

**Comment by Jarsen136** [comment]
> 
> * could we replace `round-image` class with `rounded` prop
> * I would remove "Price: None" when item is not Listed

Done

> * refactor/create new media component (reusable)

IMO it's may not that necessary for now, as the UI design is not the final version. The media elements are quite different in different tabs.

> * no more history tab?

Yes, according to the design requirement in the origin issue.

**Comment by Jarsen136** [comment]
Still work in progress
- [ ] #3843

**Comment by yangwao** [comment]
Seems to "See all" doesn't redirect to the right results? 

https://user-images.githubusercontent.com/5887929/186948073-4fa9864b-7ab3-4c78-aa6f-4e94c3b97d62.mov

**Comment by Jarsen136** [comment]
> Made issue to add skeleton items to the search
> 
> * [Add skeleton with 3-5 items in search bar  #3843](https://github.com/kodadot/nft-gallery/issues/3843)

✅ Done

Kapture 2022-08-27 at 18 50 13

Kapture 2022-08-27 at 18 53 05

**Comment by Jarsen136** [comment]
UPDATED for #3858
Kapture 2022-08-28 at 20 28 12

**Comment by yangwao** [comment]
yes!

**Comment by yangwao** [comment]
I think it's fair to go out 🤔 

let's test it through @DeluneClaire if you spot any bugs :)

**Review by preschian — APPROVED** [review]
LGTM
It looks fancy tho 🌟
need to split into smaller components in another PR

**Comment by roiLeo** [comment]
> UPDATED for #3858 Kapture 2022-08-28 at 20 28 12 

Screenshot 2022-08-29 at 08-31-15 KodaDot - Kusama NFT Market Explorer
On beta there is more space between buttons, could you take a look?

**Comment by Jarsen136** [comment]
> Screenshot 2022-08-29 at 08-31-15 KodaDot - Kusama NFT Market Explorer On beta there is more space between buttons, could you take a look?

Yes, will check it.

**Comment by DeluneClaire** [comment]
> > Screenshot 2022-08-29 at 08-31-15 KodaDot - Kusama NFT Market Explorer On beta there is more space between buttons, could you take a look?
> 
> Yes, will check it.

Did you remove the other bottom

test
?

**Comment by DeluneClaire** [comment]
On the dark mode - phone, you can't see the text.

test2

**Comment by DeluneClaire** [comment]
On the phone and the light version, the name of the NFTs you can't see, I think it should be black, not white.

test3

**Review by DeluneClaire — COMMENTED** [review]
This is a lot of changes and bugs, so maybe it would be better if someone else had an eye on it... 
But I checked: the search while
                         skeletons
                         extra line search on phone
                         option to reset reach terms
                         no scroll
                         the colours on the PC version (light/dark)
                         see all function
                         
                         And this all works for me.
                         
Then check the phone and there I find these two colours problem:
- on the light mode the name for NFTs is to light
- on the dark mode the search is too dark

File: styles/themes/_light.scss [hunk 6]
```diff
@@ -268,4 +268,16 @@ html.light-mode {
...
 		color: $primary  !important;
 	}
+
+	.main-title {
+		color: $text-light-theme;
+	}
+	.search-gallery {
+		.media-content {
+			color: gray;
+		}
```
[the remark this answers is not in the document]
Comment (Jarsen136) [reply styles/themes/_light.scss:278]: But we would like to show different color in light/dark mode, so replacing with some fixed class name was not properly. 

`has-text-xxx` may cause some style bug in dark/light mode as I just found one : )

[the remark this answers is not in the document]
Comment (roiLeo) [reply styles/themes/_light.scss:278]: `search-gallery` don't even exist? I think it's `gallery-search`

[the remark this answers is not in the document]
Comment (Jarsen136) [reply styles/themes/_light.scss:278]: yup my bad  : ) Replaced with `gallery-search`

**Comment by Jarsen136** [comment]
> Then check the phone and there I find these two colours problem:
> 
> * on the light mode the name for NFTs is to light
> * on the dark mode the search is too dark

Fixed

File: components/search/SearchResultItem.vue [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    border-radius: 4px;
+  }
+  .loading-text {
+    border-radius: 4px;
+    width: 150px;
+  }
+  .loading {
+    background: #4d5358;
+    animation: loading-keyframes 2s linear infinite;
+  }
```
[the remark this answers is not in the document]
Comment (Jarsen136) [reply components/search/SearchResultItem.vue]: OK

[the remark this answers is not in the document]
Comment (roiLeo) [reply components/search/SearchResultItem.vue]: Thank you!

File: styles/themes/_light.scss (hunk 6 above)
[the remark this answers is not in the document]
Comment (roiLeo) [reply styles/themes/_light.scss:278]: IMO, this part of code is useless 
I would change it with `has-text-grey` on `<span>` in media-content

**pic**
Screenshot 2022-08-29 at 16-29-37 KodaDot - Kusama NFT Market Explorer
Screenshot 2022-08-29 at 16-29-12 KodaDot - Kusama NFT Market Explorer

**Review by DeluneClaire — APPROVED** [review]
Checked now the colours on the phone and it works for me:)

**Comment by yangwao** [comment]
I guess it's time to merge this and do postfixes
image

- refactor into smaller components 
- working on different width on mobile
- I don't see suggestions on mobile 

image

as on beta, it has worked 
image

- Is there anything else?

**Pull request merged**